drivers/misc/ipts/ipts-companion.c

```diff
@@ -0,0 +1,203 @@
+
+	if (ipts_companion != NULL) {
+		ret = -EBUSY;
+		goto add_companion_return;
+	}
+
+	ipts_companion = companion;
+
+add_companion_return:
+
+	mutex_unlock(&ipts_companion_lock);
+	return ret;
+}
+EXPORT_SYMBOL_GPL(ipts_add_companion);
+
+int ipts_remove_companion(ipts_companion_t *companion)
+{
+	int ret = 0;
+	mutex_lock(&ipts_companion_lock);
+
+	if (ipts_companion == NULL || companion == NULL) {
+		ret = 0;
+		goto remove_companion_return;
+	}
+
+	if (ipts_companion->name != companion->name) {
+		ret = -EPERM;
+		goto remove_companion_return;
+	}
+
+	ipts_companion = NULL;
+
+remove_companion_return:
+
+	mutex_unlock(&ipts_companion_lock);
+	return ret;
+}
+EXPORT_SYMBOL_GPL(ipts_remove_companion);
+
+/*
+ * Utility functions for IPTS. These functions replace codepaths in the IPTS
+ * driver, and redirect them to the companion driver, if one was found.
+ * Otherwise the legacy code gets executed as a fallback.
+ */
+
+int ipts_request_firmware(const struct firmware **fw, const char *name,
+	struct device *device)
+{
+	int ret = 0;
+	char fw_path[MAX_IOCL_FILE_PATH_LEN];
+	mutex_lock(&ipts_companion_lock);
+
+	// Check if a companion was registered. If not, skip
+	// forward and try to load the firmware from the legacy path
+	if (ipts_companion == NULL || ipts_modparams.ignore_companion) {
+		goto request_firmware_fallback;
+	}
+
+	ret = ipts_companion->firmware_request(fw, name, device, ipts_companion);
+	if (!ret) {
+		goto request_firmware_return;
+	}
+
+request_firmware_fallback:
+
+	// If fallback loading for firmware was disabled, abort.
+	// Return -ENOENT as no firmware file was found.
+	if (ipts_modparams.ignore_fw_fallback) {
+		ret = -ENOENT;
+		goto request_firmware_return;
+	}
+
+	// No firmware was found by the companion driver, try the generic path.
+	snprintf(fw_path, MAX_IOCL_FILE_PATH_LEN, IPTS_FW_PATH_FMT, name);
+	ret = request_firmware(fw, fw_path, device);
+
+request_firmware_return:
+
+	mutex_unlock(&ipts_companion_lock);
+	return ret;
+}
+
+ipts_bin_fw_list_t *ipts_alloc_fw_list(ipts_bin_fw_info_t **fw)
+{
+	int size, len, i, j;
+	ipts_bin_fw_list_t *fw_list;
+	char *itr;
+
+	// Figure out the amount of firmware files inside of the array
+	len = 0;
+	while (fw[len] != NULL) {
+		len++;
+	}
+
+	// Determine the size that the final list will need in memory
+	size = sizeof(ipts_bin_fw_list_t);
+	for (i = 0; i < len; i++) {
+		size += sizeof(ipts_bin_fw_info_t);
+		size += sizeof(ipts_bin_data_file_info_t) *
+				fw[i]->num_of_data_files;
+	}
+
+	fw_list = kmalloc(size, GFP_KERNEL);
+	fw_list->num_of_fws = len;
+	itr = (char *)fw_list->fw_info;
+	for (i = 0; i < len; i++) {
+		*(ipts_bin_fw_info_t *)itr = *fw[i];
+		itr += sizeof(ipts_bin_fw_info_t);
+
+		for (j = 0; j < fw[i]->num_of_data_files; j++) {
+			*(ipts_bin_data_file_info_t *)itr = fw[i]->data_file[j];
+			itr += sizeof(ipts_bin_data_file_info_t);
+		}
+	}
+
+	return fw_list;
+}
+
+int ipts_request_firmware_config(ipts_info_t *ipts, ipts_bin_fw_list_t **cfg)
```

Comment:
this whole function can be written better without any gotos

Reply:
See my answer above: Not using gotos would yield duplicated `mutex_unlock` calls all over the place, and using goto in such a case is what the kernel coding guidelines advise.

Reply:
I'm not necessarily for throwing out the gotos, but I think we can reduce the code beneath the mutex. I think actually only the code until (including) `*cfg = ipts_alloc_fw_list(ipts_companion->firmware_config);` needs to be synchronized. `ipts_request_firmware` does its own checks and synchronization, and the rest is stuff on data that we own at that point.

Reply:
Also with that you'll probably get rid of one goto.

Reply:
I looked into how to reduce the amount of gotos in that function, but I found no nice way that allows for all the fallback behaviour that is there, and that doesn't require duplicating the call to `mutex_unlock`.

If you *want* me to change it, I will look into it again ofc.

Reply:
What I meant was something like this:

```c
int ipts_request_firmware_config(ipts_info_t *ipts, ipts_bin_fw_list_t **cfg)
{
	int ret = 0;
	const struct firmware *config_fw = NULL;
	mutex_lock(&ipts_companion_lock);

	// Check if a companion was registered. If not, skip
	// forward and try to load the firmware config from a file
	if (ipts_modparams.ignore_companion || ipts_companion == NULL) {
		mutex_unlock(&ipts_companion_lock);
		goto config_fallback;
	}

	if (ipts_companion->firmware_config != NULL) {
		*cfg = ipts_alloc_fw_list(ipts_companion->firmware_config);
		mutex_unlock(&ipts_companion_lock);
		return 0;
	}

config_fallback:

	// If fallback loading for the firmware config was disabled, abort.
	// Return -ENOENT as no config file was found.
	if (ipts_modparams.ignore_config_fallback) {
		ret = -ENOENT;
		goto config_return;
	}

	// No firmware config was found by the companion driver,
	// try loading it from a file now
	ret = ipts_request_firmware(&config_fw, IPTS_FW_CONFIG_FILE,
			&ipts->cldev->dev);
	if (ret)
		return ret;

	*cfg = (ipts_bin_fw_list_t *)config_fw->data;
	release_firmware(config_fw);
	return 0;
}
```

You only need the mutex to guard the direct `ipts_companion` access, the rest doesn't need to be in the critical section (especially as `ipts_request_firmware` does all that itself). I'd still keep the goto for the fallback as I don't see a cleaner option.

Reply:
Ok, yeah, that is definitly nicer. Updated the function, thank you.